[ResponseOps] Cases analytics index

adcoelho opened the pull request:
This PR is for a feature branch that is being merged into main.

The relevant PRs are:
- https://github.com/elastic/kibana/pull/219211
- https://github.com/elastic/kibana/pull/222820
- https://github.com/elastic/kibana/pull/223241
- https://github.com/elastic/kibana/pull/224388
- https://github.com/elastic/kibana/pull/224682

## Summary

This PR adds 4 new indexes with case analytics data, which are created when the cases plugin starts.

  - `.internal.cases`
  - `.internal.cases-comments`
  - `.internal.cases-attachments`
  - `.internal.cases-activity`

After the indexes are created, a backfill task for each of them is scheduled to run 1 minute after creation. This task populates the indexes with relevant data from `.kibana_alerting_cases`.

A second type of task is registered, the index synchronization task. Four of these tasks, one for each index, are scheduled to run every 5 minutes. The synchronization tasks populated the indexes with data from `.kibana_alerting_cases` that was created or updated in the last five minutes.

## How to test

You might want to start Kibana with `--verbose` to see relevant index messages in the console.

Alternatively(what I normally do), is go to `analytics_index.ts`, `backfill_task_runner.ts`, and `synchronization_task_runner.ts`, and change the `logDebug` function to call `this.logger.info` instead. This way, you will have less spam in the console.

Every log message starts with the index name between square brackets, so you can look for `[.internal.cases-` and follow what is happening.

1. You should have some existing case data, so before anything else, please create some activity, attachments, etc.
2. Add `xpack.cases.analytics.index.enabled: true` to `kibana.dev.yml` and restart Kibana.
3. Check out [this branch](https://github.com/elastic/elasticsearch/pull/129414) from the ES project.
4. Start Elastic Search with `yarn es source`. This will use the above version of Elasticsearch.
5. Wait a bit for the indexes to be created and populated(backfilled).
6. Using the dev tools:
    - Confirm the indexes exist.
    - Check the index mapping. Does it match the one in the code? Is the `_meta` field correct?
        - `x-pack/platform/plugins/shared/cases/server/cases_analytics/******_index/mappings.ts`
    - Check that the painless scripts match the code.
        - `x-pack/platform/plugins/shared/cases/server/cases_analytics/******_index/painless_scripts.ts`
    - Confirm your existing case data is in the indexes. (See **Queries** section below.)
7. Play around with cases. Some examples:
    - Create a case
    - Change status/severity
    - Attach alerts
    - Add files
    - Change category/tags
    - Add comments
    - etc
8. Go to the dev tools again and confirm all this shows up in the relevant indexes.  (See **Queries** section below.)

## Queries

In addition to the ones, below I have a few more. Things like reindexing with specific scripts or fetching relevant data from `.kibana_alerting_cases`. Ping me if you want those queries.

### Checking index content
```
GET /.internal.cases/_search
GET /.internal.cases-comments/_search
GET /.internal.cases-attachments/_search
GET /.internal.cases-activity/_search
```

### Checking index mappings
```
GET /.internal.cases
GET /.internal.cases-comments
GET /.internal.cases-attachments
GET /.internal.cases-activity
```

### Fetching the painless scripts
```
GET /_scripts/cai_cases_script_1
GET /_scripts/cai_attachments_script_1
GET /_scripts/cai_comments_script_1
GET /_scripts/cai_activity_script_1
```

### Emptying the indexes

It is sometimes useful for testing.
```
POST /.internal.cases/_delete_by_query
POST /.internal.cases-comments/_delete_by_query
POST /.internal.cases-attachments/_delete_by_query
POST /.internal.cases-activity/_delete_by_query
```

### Deleting the indexes

It is sometimes useful for testing.
```
DELETE /.internal.cases
DELETE /.internal.cases-comments
DELETE /.internal.cases-attachments
DELETE /.internal.cases-activity
```

## Release notes

Four dedicated case analytics indexes were created, allowing users to build dashboards and metrics over case data. These indexes are created on Kibana startup and updated periodically with cases, comments, attachments, and activity data.

<!--ONMERGE {"backportTargets":["8.19"]} ONMERGE-->

adcoelho committed:
[ResponseOps] Cases analytics index creation and backfill (#219211)

Closes [#221234](https://github.com/elastic/kibana/issues/221234)

**Merging into a feature branch**

## Summary

This PR adds:
- The analytics index creation when the cases plugin starts.
- Backfill task registration.
- Backfill task scheduling.
- Mapping version 1 for:
  - `.internal.cases`
  - `.internal.cases-comments`
  - `.internal.cases-attachments`
- All corresponding painless scripts.

## How to review

1. The main functionality is in `analytics_index.ts`. The class in this
file implements the index creation algorithm of the RFC.
2. Three folders instantiate this class to create the indexes we need:
  - `cases_analytics/cases_index/`
  - `cases_analytics/comments_index/`
  - `cases_analytics/attachments_index/`
3. The backfill logic can be found in
`cases_analytics/tasks/backfill_task`.
4. I moved the `retry_service` previously used by the cases connector to
a `common` folder and started using it in the `AnalyticsIndex` class.
5. The call to create the indexes is done in
`x-pack/platform/plugins/shared/cases/server/plugin.ts`.

## How to test

A few things are needed.

1. Create a role that has index creation permission. Below is what works
for me; it might be overkill.
<img width="1030" alt="Screenshot 2025-04-25 at 11 30 45"
src="https://github.com/user-attachments/assets/cbb7b384-e438-43ec-bbca-a18e62243523"
/>

2. Create a user with the role created in step 1. Let's call the user
`index_creator`. **The password should be** `changeme`.
<img width="842" alt="Screenshot 2025-04-25 at 11 31 27"
src="https://github.com/user-attachments/assets/a9547c96-086e-43b4-a442-a18f558dcb96"
/>

6. Start Kibana with this user as the Elasticsearch user - `yarn start`
with the option `--elasticsearch.username=index_creator`.

7. Kibana startup logs should show relevant messages:
```
[INFO ][plugins.cases] [.internal.cases] Checking if index exists.
[INFO ][plugins.cases] [.internal.cases-attachments] Checking if index exists.
[INFO ][plugins.cases] [.internal.cases-comments] Checking if index exists.
```

8. You can check the content of these indexes in the Dev Tools.
```
GET /.internal.cases/_search
GET /.internal.cases-comments/_search
GET /.internal.cases-attachments/_search
```

9. You can check the mapping of each of these indexes in the Dev Tools.
```
GET /.internal.cases
GET /.internal.cases-comments
GET /.internal.cases-attachments
```

10. You can check the painless scripts in the Dev Tools.
```
GET /_scripts/cai_cases_script_1
GET /_scripts/cai_attachments_script_1
GET /_scripts/cai_comments_script_1
```

---------

Co-authored-by: kibanamachine <42973632+kibanamachine@users.noreply.github.com>

elasticmachine commented:
<!-- buildkite-pr-bot-trigger-comment -->:robot: Jobs for this PR can be triggered through checkboxes. :construction: 

:information_source: To trigger the CI, please tick the checkbox below :point_down: 

- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-pull-request --> Click to trigger **kibana-pull-request** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-project-from-pr --> Click to trigger **kibana-deploy-project-from-pr** for this PR!
- [ ] <!-- buildkite-pr-bot-checkbox-trigger:kibana-deploy-cloud-from-pr --> Click to trigger **kibana-deploy-cloud-from-pr** for this PR!

adcoelho committed:
[ResponseOps] Cases analytics index synchronization (#222820)

Closes [#221232](https://github.com/elastic/kibana/issues/221232)

**Merging into a feature branch**

## Summary

This PR adds:
- Synchronization task registration.
- Synchronization task scheduling.
- Testing

## How to test

A few things are needed.

1. Create a role that has index creation permission. Below is what works
for me; it might be overkill.
<img width="1030" alt="Screenshot 2025-04-25 at 11 30 45"
src="https://github.com/user-attachments/assets/cbb7b384-e438-43ec-bbca-a18e62243523"
/>

2. Create a user with the role created in step 1. Let's call the user
`index_creator`. **The password should be** `changeme`.
<img width="842" alt="Screenshot 2025-04-25 at 11 31 27"
src="https://github.com/user-attachments/assets/a9547c96-086e-43b4-a442-a18f558dcb96"
/>

6. Start Kibana with this user as the Elasticsearch user - `yarn start`
with the option `--elasticsearch.username=index_creator`.

7.  You can check the content of these indexes in the Dev Tools.
```
GET /.internal.cases/_search
GET /.internal.cases-comments/_search
GET /.internal.cases-attachments/_search
```

8. Create Cases, Comments, and Attachments.
9. After about 5 minutes, check again the content of the indexes using
the instructions in step 7.
10. Confirm the Cases, Comments, and Attachments created in step 8 can
be seen in the analytics indexes.

adcoelho committed:
[ResponseOps] Cases analytics index configuration option (#223241)

**Merging into a feature branch**

## Summary

This PR adds a configuration option to enable the cases analytics index
logic.

`xpack.cases.analyticsIndex.enabled: true`

The task types are registered, but by default, no task is scheduled, and
no index is created.

## How to review

**The previous PRs are not merged yet, so only commits after
bd6f2b8e307fa0cb7da06358dd1836f767e3b5e1 should be reviewed.**

adcoelho committed:
[ResponseOps] Cases activity index (#224085)

**Merging into a feature branch**

## Summary

This PR adds the `.internal.cases-activity` analytics index.

## How to test

1. Add `xpack.cases.analytics.index.enabled: true` to `kibana.dev.yml`
2. Check out [this
branch](https://github.com/elastic/elasticsearch/pull/129414) from the
ES project.
3. Start Elastic Search with `yarn es source`.
4. Add a bunch of activity to your cases. Edit tags, category, etc.
5. Go to the dev tools and confirm this activity shows up in the index -
`GET /.internal.cases-activity/_search`.

elasticmachine commented:
Pinging @elastic/response-ops (Team:ResponseOps)

adcoelho committed:
Merge branch 'main' into cases-analytics-index

adcoelho committed:
[ResponseOps] Aliases for the cases analytics indexes (#224388)

**Merging into a feature branch.**

## Summary

This PR adds the following aliases when creating the cases analytics
indexes:
- `.internal.cases` -> `.cases`
- `.internal.cases-activity` -> `.cases-activity`
- `.internal.cases-attachments` -> `.cases-attachments`
- `.internal.cases-comments` -> `.cases-comments`

adcoelho committed:
[ResponseOps] Include time metrics and totals in the cases analytics index. (#224682)

## Summary

This PR adds some recently added fields to the cases analytics index.

The fields are:
- `time_to_resolve`
- `time_to_acknowledge`
- `time_to_investigate`
- `total_alerts`
- `total_comments`

adcoelho committed:
Add wait_for_completion to backfill task.

adcoelho committed:
Add timestamp to activity

cnasikas commented:
I am testing the PR with one million cases. As I test, I will post any issues/bugs I find if any. So far I found two small bugs:
- The `@timestamp` is not populated in the activity index.
- The `create_case` user action activity is copied to the activity index. Based on the painless script, this should not happen. Also, I think we should not copy them because there will be one for each case, and for the initial phase, maybe this is too much.

cnasikas commented:
@elasticmachine merge upstream

elasticmachine committed:
Merge branch 'main' into cases-analytics-index

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/311235)
* Commit: ee2a3db63388e9dae4f751d9e2d8b05d28c3ddea

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #310947](https://buildkite.com/elastic/kibana-pull-request/builds/310947) succeeded 351735780aa82dfedfd47818d30b135373d23109
* :green_heart: [Build #310725](https://buildkite.com/elastic/kibana-pull-request/builds/310725) succeeded 5d42ded2ea17601f5866be2cd312e3dbf5193be3
* :green_heart: [Build #309330](https://buildkite.com/elastic/kibana-pull-request/builds/309330) succeeded 35deb011c3ca331e20ea39463fb2360f7233a5f2

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311235","number":311235,"commit":"ee2a3db63388e9dae4f751d9e2d8b05d28c3ddea"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310947","number":310947,"commit":"351735780aa82dfedfd47818d30b135373d23109"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310725","number":310725,"commit":"5d42ded2ea17601f5866be2cd312e3dbf5193be3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309330","number":309330,"commit":"35deb011c3ca331e20ea39463fb2360f7233a5f2"}],"number":311235}
buildkite-pr-comment-->

cnasikas reviewed (commented):
I tested documents with no data at all and I found some small issues in the scripts. I left some comments on the script files. Also, I noticed that when the reindex fails, due to painless script errors, it updates the last success time to be the last attempt even though an error happened. I noticed this because the next time it ran it skipped the case with the corrupted data that was causing the script to fail. Lastly, the files (attachments) are not indexed at all.

cnasikas committed:
Fix bug with errors in reindex operations

cnasikas commented:
> Also, I noticed that when the reindex fails, due to painless script errors, it updates the last success time to be the last attempt even though an error happened.

Fixed in https://github.com/elastic/kibana/pull/223405/commits/947cd802a19016584ead580d8b75543ab8740fac

adcoelho committed:
Addressing PR comments.

adcoelho committed:
[ResponseOps] Updated activity index (#224875)

## Summary

The cases activity analytics index will only have user actions of type:
- tags
- status
- severity
- category

adcoelho commented:
> The @timestamp is not populated in the activity index.

Fixed.

> The create_case user action activity is copied to the activity index.

Fixed in [27dc400](https://github.com/elastic/kibana/pull/223405/commits/27dc400b12ad9906b8135c6253c16b9c8281142c) by whitelisting the types we want to see.

**I also changed the default of the config var to `true`**

elasticmachine commented:
## :green_heart: Build Succeeded
* [Buildkite Build](https://buildkite.com/elastic/kibana-pull-request/builds/311684)
* Commit: 27dc400b12ad9906b8135c6253c16b9c8281142c

### Metrics [[docs](https://www.elastic.co/guide/en/kibana/master/ci-metrics.html)]
> ✅ unchanged

### History
* :green_heart: [Build #311606](https://buildkite.com/elastic/kibana-pull-request/builds/311606) succeeded 947cd802a19016584ead580d8b75543ab8740fac
* :green_heart: [Build #311235](https://buildkite.com/elastic/kibana-pull-request/builds/311235) succeeded ee2a3db63388e9dae4f751d9e2d8b05d28c3ddea
* :green_heart: [Build #310947](https://buildkite.com/elastic/kibana-pull-request/builds/310947) succeeded 351735780aa82dfedfd47818d30b135373d23109
* :green_heart: [Build #310725](https://buildkite.com/elastic/kibana-pull-request/builds/310725) succeeded 5d42ded2ea17601f5866be2cd312e3dbf5193be3
* :green_heart: [Build #309330](https://buildkite.com/elastic/kibana-pull-request/builds/309330) succeeded 35deb011c3ca331e20ea39463fb2360f7233a5f2

cc @adcoelho

<!--buildkite-pr-comment-kibana-pull-request
{"builds":[{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311684","number":311684,"commit":"27dc400b12ad9906b8135c6253c16b9c8281142c"},{"buildStatus":{"state":"passed","success":true,"hasRetries":false,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311606","number":311606,"commit":"947cd802a19016584ead580d8b75543ab8740fac"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/311235","number":311235,"commit":"ee2a3db63388e9dae4f751d9e2d8b05d28c3ddea"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310947","number":310947,"commit":"351735780aa82dfedfd47818d30b135373d23109"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/310725","number":310725,"commit":"5d42ded2ea17601f5866be2cd312e3dbf5193be3"},{"buildStatus":{"state":"passed","success":true,"hasRetries":true,"hasNonPreemptionRetries":false},"url":"https://buildkite.com/elastic/kibana-pull-request/builds/309330","number":309330,"commit":"35deb011c3ca331e20ea39463fb2360f7233a5f2"}],"number":311684}
buildkite-pr-comment-->

cnasikas reviewed (approved):
Superb work 🎉 !!

kibanamachine commented:
## 💔 All backports failed

| Status | Branch | Result |
|:------:|:------:|:------|
|❌|8.19|Backport failed because of merge conflicts|

### Manual backport
To create the backport manually run:
```
node scripts/backport --pr 223405
```

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 9.6.6 -->

adcoelho commented:
## 💚 All backports created successfully

| Status | Branch | Result |
|:------:|:------:|:------|
|✅|8.19|[<img src="https://img.shields.io/github/pulls/detail/state/elastic/kibana/225019">](https://github.com/elastic/kibana/pull/225019)|

Note: Successful backport PRs will be merged automatically after passing CI.

### Questions ?
Please refer to the [Backport tool documentation](https://github.com/sorenlouv/backport)

<!--- Backport version: 10.0.1 -->
